tests/numerics/fv/test_tpsa.py

```diff
@@ -0,0 +1,1458 @@
+                sps.csr_array((g.num_faces, n_rot_face)),
+            ],
+        ]
+    )
+
+    div = sps.block_diag(
+        [
+            pp.fvutils.vector_divergence(g),
+            div_rot,
+            pp.fvutils.scalar_divergence(g),
+        ],
+        format="csr",
+    )
+
+    accum = sps.block_diag(
+        [
+            sps.csr_array((g.num_cells * g.dim, g.num_cells * g.dim)),
+            sps.eye(n_rot_cell),
+            sps.eye(g.num_cells),
+        ],
+        format="csr",
+    )
+
+    return flux, rhs_matrix, div, accum
+
+
+def _solve(
+    flux: sps.sparray,
+    rhs_matrix: sps.sparray,
+    div: sps.sparray,
+    accum: sps.sparray,
+    bound_vec: np.ndarray,
+) -> np.ndarray:
+    """Solve the TPSA problem.
+
+    Parameters:
+        flux: Discretization of the face terms as a block matrix.
+        rhs_matrix: Discretization of the boundary conditions.
+        div: Divergence matrix for the face terms.
+        accum: Accumulation matrix for the cell center terms.
+        bound_vec: Array of boundary condition values.
+
+    Returns:
+        np.ndarray: Array of cell center values.
+    """
+    b = -div @ rhs_matrix @ bound_vec
+
+    # Assemble and solve. The minus sign on accum follows from the definition of the
+    # governing equations in the paper.
+    A = div @ flux - accum
+    x = sps.linalg.spsolve(A, b)
+    return x
+
+
+def _set_bc_by_direction(
+    g: pp.Grid,
+    d: dict,
+    type_south: Literal["dir", "neu"],
+    type_east: Literal["dir", "neu"],
+    type_north: Literal["dir", "neu"],
+    type_west: Literal["dir", "neu"],
+    type_bottom: Optional[Literal["dir", "neu"]] = None,
+    type_top: Optional[Literal["dir", "neu"]] = None,
+) -> np.ndarray:
+    """Set the boundary conditions on the grid, based on the types of boundary conditions
+    given.
+
+    The boundary condition has value 0.1 on the south face, -0.1 on the east face, and
+    0.1 on the bottom face (if 3d). The remaining faces have homogeneous conditions.
+    The conditions correspond to a compressive force on the south, east and (if
+    relevant) bottom faces.
+
+    Parameters:
+        g: Grid object.
+        d: Dictionary of parameters.
+        type_south: Type of boundary condition on the south face.
+        type_east: Type of boundary condition on the east face.
+        type_north: Type of boundary condition on the north face.
+        type_west: Type of boundary condition on the west face.
+        type_bottom: Type of boundary condition on the bottom face. Only relevant for 3d
+            grids.
+        type_top: Type of boundary condition on the top face. Only relevant for 3d
+            grids.
+
+    Returns:
+        np.ndarray: Array of boundary condition values, as a g.dim x num_faces array.
+
+    """
+
+    face_ind = g.get_all_boundary_faces()
+    nf = face_ind.size
+
+    # Find the faces on the boundary in each direction
+    min_coord, max_coord = pp.domain.grid_minmax_coordinates(g)
```

Comment:
Please consider using function in test_mpsa (moving it to applications?)

Reply:
Good point. Will make a function in pp.geometry.domains.